File: hell/workspace/kritsn/backend/microservice-exp/lib/src/main/kotlin/com/kritsn/lib/util/Timber.java
package com.kritsn.lib.util;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Copyright © 2025 Kritsn LLP. All rights reserved.
 *
 * @author Radhey (hr-sh)
 * @since Aug 13, 2025
 */

public class Timber {

    // Automatically get the caller's class name as the log tag
    private static String getTag() {
        StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();

        for (StackTraceElement element : stackTrace) {
            String className = element.getClassName();

            // Skip this logger class and Java internal classes
            if (!className.equals(Timber.class.getName())
                    && !className.startsWith("java.")
                    && !className.startsWith("sun.")) {
                // Only return the simple class name, not the full package
                return className.substring(className.lastIndexOf('.') + 1);
            }
        }
        return "UnknownClass";
    }

    public static void d(String message) {
        Logger logger = LoggerFactory.getLogger(getTag());
        logger.debug(message);
    }

    public static void i(String message) {
        Logger logger = LoggerFactory.getLogger(getTag());
        logger.info(message);
    }

    public static void w(String message) {
        Logger logger = LoggerFactory.getLogger(getTag());
        logger.warn(message);
    }

    public static void e(String message) {
        Logger logger = LoggerFactory.getLogger(getTag());
        logger.error(message);
    }

    public static void e(String message, Throwable throwable) {
        Logger logger = LoggerFactory.getLogger(getTag());
        logger.error(message, throwable);
    }
}

File: hell/workspace/kritsn/backend/microservice-exp/lib/src/main/kotlin/com/kritsn/lib/logger/Timber.kt
package com.kritsn.lib.logger

import org.slf4j.LoggerFactory
import org.slf4j.spi.LocationAwareLogger
import kotlin.jvm.java

/**
 * Copyright © 2025 Kritsn LLP. All rights reserved.
 *
 * @author Radhey (hr-sh)
 * @since Sep 08, 2025
 */

object Timber {
    @PublishedApi internal val base = LoggerFactory.getLogger("app")
    @PublishedApi internal val la = base as? LocationAwareLogger
    @PublishedApi internal val FQCN = Timber::class.java.name
//    @PublishedApi internal val tag = Throwable

    inline fun v(crossinline m: () -> String) = {
        log(LocationAwareLogger.TRACE_INT, null, m)
    }
    inline fun d(crossinline m: () -> String) = log(LocationAwareLogger.DEBUG_INT, null, m)
    inline fun i(crossinline m: () -> String) = log(LocationAwareLogger.INFO_INT,  null, m)
    inline fun w(crossinline m: () -> String) = log(LocationAwareLogger.WARN_INT,  null, m)
    inline fun e(crossinline m: () -> String) = log(LocationAwareLogger.ERROR_INT, null, m)
    inline fun e(t: Throwable, crossinline m: () -> String) =
        log(LocationAwareLogger.ERROR_INT, t, m)

    @PublishedApi internal inline fun log(level: Int, t: Throwable?, crossinline m: () -> String) {
        if (!isEnabled(level)) return
        val msg = m()
        la?.log(null, FQCN, level, msg, null, t) ?: when (level) {
            LocationAwareLogger.TRACE_INT -> if (t != null) base.trace(msg, t) else base.trace(msg)
            LocationAwareLogger.DEBUG_INT -> if (t != null) base.debug(msg, t) else base.debug(msg)
            LocationAwareLogger.INFO_INT  -> if (t != null) base.info (msg, t) else base.info (msg)
            LocationAwareLogger.WARN_INT  -> if (t != null) base.warn (msg, t) else base.warn (msg)
            else                          -> if (t != null) base.error(msg, t) else base.error(msg)
        }
    }
    @PublishedApi internal fun isEnabled(level: Int) = when (level) {
        LocationAwareLogger.TRACE_INT -> base.isTraceEnabled
        LocationAwareLogger.DEBUG_INT -> base.isDebugEnabled
        LocationAwareLogger.INFO_INT  -> base.isInfoEnabled
        LocationAwareLogger.WARN_INT  -> base.isWarnEnabled
        else                          -> base.isErrorEnabled
    }
}